src/mo_frontend/typing.ml

```diff
@@ -1379,6 +1385,24 @@ and check_exp' env0 t exp : T.typ =
       warn env exp.at "M0155" "operator may trap for inferred type%a"
         display_typ_expand t;
     t
+  | ToCandidE (ot, exps), _ ->
+    begin match t with
+    | T.Prim T.Blob -> begin
+        let ts = List.map (infer_exp env) exps in
+        ot := T.Tup ts;
+        T.Prim T.Blob
+      end
+    | _ ->
+      error env exp.at "MXXXX" "to_candid produces Blob, not type%a"
+        display_typ_expand t
+    end
+  | FromCandidE (ot, exp1), t when T.shared t && T.is_opt t ->
```

Comment:
```suggestion
  | FromCandidE (ot, exp1), t when T.shared t && T.is_opt t ->
```
Can you assume `t` is normalized here? If not, you need to `normalize` (or perhaps even `promote`) before checking `T.is_opt...` Maybe look if you can find some similar code elsewhere

Reply:
Actually, I think you can assume it's normalized here. Nvm,